File: sorting algorithms/quick sort/quickSort.py
# This function takes last element as pivot, places
# the pivot element at its correct position in sorted
# array, and places all smaller (smaller than pivot)
# to left of pivot and all greater elements to right
# of pivot
def partition(arr, low, high):
	i = (low-1)		 # index of smaller element
	pivot = arr[high]	 # pivot

	for j in range(low, high):

		# If current element is smaller than or
		# equal to pivot
		if arr[j] <= pivot:

			# increment index of smaller element
			i = i+1
			arr[i], arr[j] = arr[j], arr[i]

	arr[i+1], arr[high] = arr[high], arr[i+1]
	return (i+1)



# The main function that implements QuickSort
# arr[] --> Array to be sorted,
# low --> Starting index,
# high --> Ending index
# Function to do Quick sort
def quickSort(arr, low, high):
	if len(arr) == 1:
		return arr
	if low < high:

		# pi is partitioning index, arr[p] is now
		# at right place
		pi = partition(arr, low, high)

		# Separately sort elements before
		# partition and after partition
		quickSort(arr, low, pi-1)
		quickSort(arr, pi+1, high)


def printArr(arr):
    for element in arr:
        print(element,end=" ")
    print("")


# Driver code to test above
n = int(input("Enter the size of the list "))
print("\n")
arr = list(int(num) for num in input("Enter elements ").strip().split())[:n]
print("Unsorted List: ")
printArr(arr)
quickSort(arr, 0, n-1)
print("\n")
printArr(arr)

File: linked-list/simple-linked-list/java implementation/QuickSort.java
public class QuickSort {
    static class Node{
        int data;
        Node next;
    }
    static Node newNode(int d){
        Node n = new Node();
        n.data = d;
        n.next = null;
        return n;
    } 
    static void print(Node node){
        while(node!=null){
            System.out.print(node.data + " ");
            node = node.next;
        }
    } 
    static void append(Node head, int d){
        Node n = newNode(d);
        if(head==null){
            head = newNode(d);
            return;
        }
        n.next = null;
        Node last = head;
        while(last.next!=null)
            last = last.next;
        last.next = n;
    }
    static Node partition(Node startNode, Node endNode ){
        if(startNode == endNode || startNode == null || endNode == null)
            return startNode;
        Node pivot_prev = startNode;
        Node curr = startNode;
        int pivot = endNode.data;
        while(startNode!=endNode){
            if(startNode.data<pivot){
                pivot_prev = curr;
                int temp = curr.data;
                curr.data = startNode.data;
                startNode.data = temp;
                curr = curr.next;
            }
            startNode = startNode.next;
        }
        int temp = curr.data;
        curr.data = pivot;
        endNode.data = temp;

        return pivot_prev;
    }
    static void sort(Node left,Node right){
        if(left == null || right == left || left == right.next)
            return;
        Node pivotPrev = partition(left,right);
        sort(left,pivotPrev);
        if (pivotPrev != null && pivotPrev == left)
        sort(pivotPrev.next, right);
        else if(pivotPrev != null && pivotPrev.next != null)
            sort(pivotPrev.next.next,right); 
    }
    public static void main(String[] args) {
        Node head = newNode(90);
        head.next = newNode(34);
        head.next.next = newNode(45);
        head.next.next.next = newNode(20);
        head.next.next.next.next = newNode(74);
        Node n =  head;
        while(n.next!=null)
            n = n.next;
        System.out.println("Linked list: ");
        print(head);
        System.out.println();
        System.out.println("Linked list after sorting: ");
        sort(head,n);
        print(head);

    }

}
